### PR TITLE
ISSUE-367 Contacts webadmin routes

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

**Comment by quantranhong1999** [comment]
Can you add a commit that binding guice module? (so I can test with memory image)

**Review by vttranlina — APPROVED** [review]

**Comment by quantranhong1999** [comment]
I think I spotted a contact webadmin routes bug while unit testing the LSC script: Get contact details should return 404 when no contact exists.
Reproduce test:
```java
        @Test
        void getContactShouldNotFoundWhenNoContactExists() throws Exception {
            ContactFields contactFields = new ContactFields(new MailAddress(mailAddressA), firstnameA, surnameA);
            Mono.from(emailAddressContactSearchEngine.index(CONTACT_DOMAIN, contactFields)).block();

            String response = given()
                .get("/" + "nonExistingLocalPart").prettyPeek()
                .then()
                .statusCode(NOT_FOUND_404)
                .contentType(JSON)
                .extract()
                .body()
                .asString();;
        }
```

**Comment by Arsnael** [comment]
I considered returning empty body when contact does not exist but I guess you are right. Will return 404 instead, makes more sense

File: tmail-backend/webadmin/webadmin-email-address-contact/src/test/java/com/linagora/tmail/webadmin/EmailAddressContactRoutesTest.java [hunk 1]
```diff
@@ -524,4 +524,141 @@ void deleteContactShouldOnlyUseMailAddressLocalPart() {
...
+            .when()
+                .put("/" + contactFields.address().getLocalPart())
+            .then()
+                .statusCode(NO_CONTENT_204);
+
+            assertThat(Flux.from(emailAddressContactSearchEngine.list(CONTACT_DOMAIN))
+                    .map(EmailAddressContact::fields)
+                    .collectList()
+                    .block())
+                .containsOnly(new ContactFields(new MailAddress(mailAddressA), "Bobby", ""));
```
Comment (quantranhong1999) [inline tmail-backend/webadmin/webadmin-email-address-contact/src/test/java/com/linagora/tmail/webadmin/EmailAddressContactRoutesTest.java:619]: Can we keep the surname/firstname when the according input field is null?

I am afraid that when the is an UPDATE OBJECT on LSC not related to contact (for other events like update mailAlias,  we can miss the firstname/surname field) => We would reset the names while we do not intend to.
E.g:
```
mainid: user@james.org
operation: UPDATE_OBJECT
sources: REPLACE_VALUES
 - alias-to-user@james.org
 - alias-to-user_bis@james.org
```
Another E.g:
```
mainid: user@james.org
operation: UPDATE_OBJECT
firstname: REPLACE_VALUES
 - Bobby
```
Here we just want to update firstname => With this current API, we would be unintended to reset surname also.

Not sure though, what do you guys think? LSC expert @vttranlina ?

Comment (vttranlina) [reply tmail-backend/webadmin/webadmin-email-address-contact/src/test/java/com/linagora/tmail/webadmin/EmailAddressContactRoutesTest.java:619]: It is an edge case, I don't think in real life we have a case remove firstname, surname to empty. 
So we can ignore updating the  firstname/surname when null

Comment (Arsnael) [reply tmail-backend/webadmin/webadmin-email-address-contact/src/test/java/com/linagora/tmail/webadmin/EmailAddressContactRoutesTest.java:619]: That would quite complexify the code TMail side honestly... It means if some fields are missing I would need to go fetch the contact details first, and then fill up missing first or surname with what's in the contact...

So... LSC does a kind of patch then for updates? (partial updates)

Comment (quantranhong1999) [reply tmail-backend/webadmin/webadmin-email-address-contact/src/test/java/com/linagora/tmail/webadmin/EmailAddressContactRoutesTest.java:619]: > LSC does a kind of patch then for updates?

It's a rare case. But it is possible I think. So the question here is should we support that partial update? If not we may encounter missing data when updating partial from the LSC side.

Comment (Arsnael) [reply tmail-backend/webadmin/webadmin-email-address-contact/src/test/java/com/linagora/tmail/webadmin/EmailAddressContactRoutesTest.java:619]: Ok I will try to take a look...

Comment (Arsnael) [reply tmail-backend/webadmin/webadmin-email-address-contact/src/test/java/com/linagora/tmail/webadmin/EmailAddressContactRoutesTest.java]: Let me know if that answers your concern :)

Comment (quantranhong1999) [reply tmail-backend/webadmin/webadmin-email-address-contact/src/test/java/com/linagora/tmail/webadmin/EmailAddressContactRoutesTest.java]: I think it does. I will check against LSC. Thank you :')

File: tmail-backend/webadmin/webadmin-email-address-contact/src/main/java/com/linagora/tmail/webadmin/EmailAddressContactRoutes.java [hunk 2]
```diff
@@ -187,6 +188,27 @@ public Route updateContact() {
...
+
+    private ContactFields fillUp(ContactFields presentFields, ContactFields updatedFields) {
+        String firstname = updatedFields.firstname();
+        String surname = updatedFields.surname();
+        if (firstname.isEmpty()) {
+            firstname = presentFields.firstname();
+        }
+        if (surname.isEmpty()) {
+            surname = presentFields.surname();
+        }
```
Comment (quantranhong1999) [inline tmail-backend/webadmin/webadmin-email-address-contact/src/main/java/com/linagora/tmail/webadmin/EmailAddressContactRoutes.java:208]: Sorry to bother you again. I think we should not update names when the field is null but should update when it is empty.
E.g: null cases
```
{
  "firstname": "a firstname"
}
```
or 
```
{
	"firstname": "a firstname",
	"surname": null
}
```
IMO if we use Optional we can easily parse this. Optional.empty would be diff from an empty string like this (which produce Optional.of(""):
```
{
	"firstname": "a firstname",
	"surname": ""
}
```

Comment (Arsnael) [reply tmail-backend/webadmin/webadmin-email-address-contact/src/main/java/com/linagora/tmail/webadmin/EmailAddressContactRoutes.java:208]: Good point, I omitted that case. Will solve it monday

File: tmail-backend/jmap/extensions-es-v7/src/main/java/com/linagora/tmail/james/jmap/ESEmailAddressContactSearchEngine.java [hunk 3]
```diff
@@ -120,6 +124,76 @@ public Publisher<EmailAddressContact> autoComplete(AccountId accountId, String p
...
+        SearchRequest request = new SearchRequest(configuration.getDomainContactReadAliasName().getValue())
+            .source(new SearchSourceBuilder()
+                .query(QueryBuilders.boolQuery()
+                    .should(QueryBuilders.termQuery(DOMAIN, domain.asString()))
+                    .minimumShouldMatch(1)));
+
+        return client.search(request, RequestOptions.DEFAULT)
+            .flatMapIterable(searchResponse -> ImmutableList.copyOf(searchResponse.getHits().getHits()))
+            .map(Throwing.function(this::extractContentFromHit).sneakyThrow());
+    }
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions-es-v7/src/main/java/com/linagora/tmail/james/jmap/ESEmailAddressContactSearchEngine.java:151]: To do a full listingm a simple search request is not enough as the count of elements is bounded.

You need to have a scrolled search here? Maybe we could have a test trying to list 1000+ contacts...

File: tmail-backend/jmap/extensions-es-v7/src/main/java/com/linagora/tmail/james/jmap/ESEmailAddressContactSearchEngine.java [hunk 4]
```diff
@@ -120,6 +124,76 @@ public Publisher<EmailAddressContact> autoComplete(AccountId accountId, String p
...
+    private EmailAddressContact extractContactFromSource(Map<String, Object> source) throws AddressException {
+        return new EmailAddressContact(
+            UUID.fromString((String) source.get(CONTACT_ID)),
+            new ContactFields(
+                new MailAddress((String) source.get(EMAIL)),
+                (String) source.get(FIRSTNAME),
+                (String) source.get(SURNAME)
+            ));
+    }
+
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions-es-v7/src/main/java/com/linagora/tmail/james/jmap/ESEmailAddressContactSearchEngine.java:196]: Outside of this range but please avoid duplicating calls to `getSourceAsMap` within `extractContentFromHit`.

Do we have tests for contacts autocomplete without FIRSTNAME SURNAME ?

File: tmail-backend/webadmin/webadmin-email-address-contact/src/test/java/com/linagora/tmail/webadmin/EmailAddressContactRoutesTest.java [hunk 5]
```diff
@@ -0,0 +1,865 @@
...
+        domainList.configure(DomainListConfiguration.DEFAULT);
+        domainList.addDomain(CONTACT_DOMAIN);
+
+        emailAddressContactSearchEngine = new InMemoryEmailAddressContactSearchEngine();
+
+        EmailAddressContactRoutes routes = new EmailAddressContactRoutes(emailAddressContactSearchEngine, domainList, new JsonTransformer());
+        webAdminServer = WebAdminUtils.createWebAdminServer(routes).start();
+
+        RestAssured.requestSpecification = WebAdminUtils.buildRequestSpecification(webAdminServer)
+            .setBasePath(String.format(BASE_PATH, CONTACT_DOMAIN.asString()))
```
Comment (chibenwa) [inline tmail-backend/webadmin/webadmin-email-address-contact/src/test/java/com/linagora/tmail/webadmin/EmailAddressContactRoutesTest.java:81]: Please avoid using those production constants in tests and write a string `"/path/to/a"` 

 -> More explicit
 -> Test will break if path is changed. Which we want, as people will realize this is a blocking change.

Comment (Arsnael) [reply tmail-backend/webadmin/webadmin-email-address-contact/src/test/java/com/linagora/tmail/webadmin/EmailAddressContactRoutesTest.java:81]: That's a local constant to the test, it's not the one declared in prod code. But I get it's confusing, will fix!

File: tmail-backend/jmap/extensions-es-v7/src/main/java/com/linagora/tmail/james/jmap/ESEmailAddressContactSearchEngine.java (hunk 4 above)
Comment (Arsnael) [reply tmail-backend/jmap/extensions-es-v7/src/main/java/com/linagora/tmail/james/jmap/ESEmailAddressContactSearchEngine.java:196]: > Do we have tests for contacts autocomplete without FIRSTNAME SURNAME ?

There is unit tests regarding it. Do you want some for the jmap method tests as well or that's enough?

**Review by chibenwa — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Comment by Arsnael** [comment]
Squashed fixups
